src/internal/pjs/server_test.go

```diff
@@ -98,11 +98,134 @@ func TestRunJob(t *testing.T) {
 		// for now we do nothing here, simply return the input filesets
 		resp.Input = in.Input
 		return nil
-	})
+	}, nil)
 	require.NoError(t, err)
 	require.Equal(t, in.Input, out.Output)
 }
 
+func TestCancelJob(t *testing.T) {
+	t.Run("cancel job before processing", func(t *testing.T) {
+		ctx := pctx.TestContext(t)
+		c, fc := setupTest(t)
+		resp := createJob(t, ctx, c, fc)
+		jid := resp.Id.Id
+		_, err := c.CancelJob(ctx, &pjs.CancelJobRequest{
+			Job: &pjs.Job{
+				Id: jid,
+			},
+		})
+		require.NoError(t, err)
+		inspectJobResp, err := c.InspectJob(ctx, &pjs.InspectJobRequest{
+			Job: &pjs.Job{
+				Id: jid,
+			},
+		})
+		require.NoError(t, err)
+		require.NotNil(t, inspectJobResp.Details)
+		jobInfo := inspectJobResp.Details.JobInfo
+		require.Equal(t, pjs.JobState_DONE, jobInfo.State)
+		// todo: the queue should be empty. This can be tested when inspectQueue is ready
+	})
+	t.Run("cancel a processing job", func(t *testing.T) {
+		ctx := pctx.TestContext(t)
+		c, fc := setupTest(t)
+
+		program := createFileSet(t, fc, map[string][]byte{"program": []byte("foo")})
+		createResp := createJob(t, ctx, c, fc, func(req *pjs.CreateJobRequest) {
+			req.Program = program
+		})
+
+		afterDequeue := func() {
```

Comment:
It looks like afterDequeue is always executed right before `fn`. On thing you could do is just call these steps within fn(), if you want to keep them logically separated (because you want fn to act like its simulating a job), then its probably better to keep them separated.

Reply:
Oh yea, afterDequeue is only used once for that test case. I'll remove this and call cancel just inside `fn`()